task/tkn-bundle/0.2/tkn-bundle.yaml

```diff
@@ -132,106 +132,292 @@ spec:
-
-        # If task is kustomized, generate it and replace the FILES value with the value of the generated task
-        # All other files will be ignored.
-        if is_kustomized_task "$task_dir"; then
-            echo "Generating a kustomized task from $task_dir"
-            FILES=("$(generate_kustomized_task "$task_dir")")
-        else
-            echo "Task is not Kustomized - continue"
+
+            local prev_bundle_digest
+            prev_bundle_digest=$(jq -r ".annotations.\"${ANNOTATION_PREVIOUS_MIGRATION_BUNDLE}\"" <"${HOME}/manifest.json")
+            if [ -n "$prev_bundle_digest" ] && [ "$prev_bundle_digest" != "null" ]; then
+                # This bundle points to a previous bundle that has migration.
+                echo "$prev_bundle_digest"
+                return 0
+            fi
+            # Otherwise, output empty to indicate not point any bundle.
+            echo
+            return 0
+          done
+          return 0
+      }
+
+      # check if migration file exists in given task bundle and task_version.
+      # Arguments: task_bundle, task_version, migration_file
+      check_existing_migration_file() {
+        local -r task_bundle=$1
+        local -r task_version=$2
+        local -r migration_file=$3
+
+        # Check if task bundle has an attached migration file.
+        local filename
+        local found=
+        local artifact_refs
+
+        # List attached artifacts, that have specific artifact type and annotation.
+        # Then, find out the migration artifact.
+        #
+        # Minimum version oras 1.2.0 is required for option --format
+        artifact_refs=$(
+          oras discover "$task_bundle" --artifact-type "$ARTIFACT_TYPE_TEXT_XSHELLSCRIPT" --format json | \
+          jq -r "
+            .manifests[]
+            | select(.annotations.\"${ANNOTATION_IS_MIGRATION}\" == \"true\")
+            | .reference"
+        )
+        while read -r artifact_ref; do
+          if [ -z "$artifact_ref" ]; then
+              continue
+          fi
+          filename=$(
+            retry oras pull --format json "$artifact_ref" | jq -r "
+              .files[]
+              | select(.annotations.\"org.opencontainers.image.title\" == \"${task_version}.sh\")
+              | .annotations.\"org.opencontainers.image.title\"
+              "
+          )
+
+          if [ -n "$filename" ]; then
+            if diff "$filename" "$migration_file" >/dev/null; then
+              found=true
+              break
+            else
+              echo "error: task bundle $task_bundle has migration artifact $artifact_ref, but the migration content is different: $filename" 1>&2
+              exit 1
+            fi
+          fi
+        done <<<"$artifact_refs"
+
+        # the same migration file exist
+        if [ "$found" == "true" ]; then
+          echo 1
         fi
+        # migratioan file doesn't exist
+        echo 0
+      }
```

Comment:
This function mixes the meaning of non-zero exit status. 1 is originally used to indicate an error, whereas 1 is also used to indicate a bundle has the migration, that implies the function succeeds.

Suggested update:

* rename `check_existing_migration_file` to some other names like `bundle_has_migration`, which implies function responds a boolean result.
* using 0 and non-zero to indicate whether function succeeds or not
* output `true` and `false` to indicate whether a bundle has the migration.